### PR TITLE
KBC-2383 branch metadata create

### PR DESCRIPTION
**Pull request by martinjunger**
Jira: KBC-2383

Before asking for review make sure that:

- [x] New client method(s) has tests
- [x] Apiary file is updated
- [x] You declared if there is a BC break or not (will affect next release of a client)

---

### CODE REVIEW

File: tests/Common/BranchMetadataTest.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+        // list metadata
+        $metadata = $defaultMdClient->listBranchMetadata();
+        // TODO change after DELETE is ready
+        //$this->assertCount(2, $metadata);
+        $this->assertIsArray($metadata);
+
+        // add metadata
+        /** @var array $metadata */
+        $metadata = $defaultMdClient->postBranchMetadata(self::TEST_METADATA);
```
Comment (tomasfejfar) [inline tests/Common/BranchMetadataTest.php:64]: nemělo by být "addBranchMetadata"?

Comment (tomasfejfar) [reply tests/Common/BranchMetadataTest.php:64]: IMHO v metodách máme add/remove ne post/delete

File: tests/Common/BranchMetadataTest.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+     * @return void
+     */
+    private function assertMetadataEquals(array $expected, array $actual)
+    {
+        foreach ($expected as $key => $value) {
+            self::assertArrayHasKey($key, $actual);
+            self::assertSame($value, $actual[$key]);
+        }
+        self::assertArrayHasKey('timestamp', $actual);
+    }
```
Comment (tomasfejfar) [inline tests/Common/BranchMetadataTest.php:153]: Tohle bude nejspíš i u těch config metadat, tak by stálo za to to vytahnout do traitu

**Review by tomasfejfar — CHANGES_REQUESTED** [review]

File: tests/Common/BranchMetadataTest.php (hunk 1 above)
Comment (martinjunger) [reply tests/Common/BranchMetadataTest.php:64]: jo, to by asi mohlo

File: apiary.apib [hunk 3]
```diff
@@ -5146,6 +5146,80 @@ Deletes the specified metadata item from the component configuration.
...
+All requests must include the branch (`/branch/{branchId}`) in the URL. This is also true for default (production) branch.
+You can find out the id of your default branch by calling [List Branches](#reference/development-branches/branches/list-branches) and finding branch with  attribute `isDefault` set to `true`.
+Or you can use placeholder `default` to use with project's default branch.
+
+Note: implemented only for default branch!
+
+### Create or Update [POST]
+Sets multiple metadata with one call. If the given key and provider combination already exist
+for the branch, the data will be updated with the new value and timestamp. All metadata associated
+with the branch are returned to the response.
```
[the remark this answers is not in the document]
Comment (martinjunger) [reply apiary.apib:5161]: hadat se s tebou nebudu, ale LanguageTool tvrdi neco jinyho:
image

File: apiary.apib [hunk 4]
```diff
@@ -5146,6 +5146,80 @@ Deletes the specified metadata item from the component configuration.
...
+                {
+                    "id": "19509",
+                    "key": "some-key",
+                    "value": "Some value",
+                    "timestamp": "2017-02-13 23:36:10"
+                }
+            ]
+
+### List [GET]
+Metadata can be listed for your branch.
```
[the remark this answers is not in the document]
Comment (martinjunger) [reply apiary.apib:5195]: acho jo, ty cleny...

File: src/Keboola/StorageApi/DevBranchesMetadata.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+     * @return mixed|string
+     */
+    public function listBranchMetadata()
+    {
+        return $this->client->apiGet("metadata");
+    }
+
+    /**
+     * @param array $metadata
+     * @return mixed|string
```
[the remark this answers is not in the document]
Comment (martinjunger) [reply src/Keboola/StorageApi/DevBranchesMetadata.php:26]: no, kdyz to vrati 500, tak dostanes HTML kod, tak mozna proto i string...

[the remark this answers is not in the document]
Comment (martinjunger) [reply src/Keboola/StorageApi/DevBranchesMetadata.php:26]: ale jo, asi bude lepsi vracet rovnou array... i kdyz tohle `mixed|string` je snad u vsech API metod

File: tests/Common/BranchMetadataTest.php (hunk 1 above)
Comment (martinjunger) [reply tests/Common/BranchMetadataTest.php:64]: upraveno

File: src/Keboola/StorageApi/DevBranchesMetadata.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+     * @var BranchAwareClient
+     */
+    private $client;
+
+    public function __construct(BranchAwareClient $client)
+    {
+        $this->client = $client;
+    }
+    /**
+     * @return mixed|string
```
[the remark this answers is not in the document]
Comment (martinjunger) [reply src/Keboola/StorageApi/DevBranchesMetadata.php:17]: upraveno na array

File: apiary.apib (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply apiary.apib:5161]: A nebo to napsat bez toho response. 
```suggestion
with the branch are returned.
```

File: tests/Common/BranchMetadataTest.php (hunk 2 above)
Comment (martinjunger) [reply tests/Common/BranchMetadataTest.php:153]: done

**Review by tomasfejfar — APPROVED** [review]

**Pull request merged**